src/core/adapters/votingEscrow.ts

```diff
@@ -0,0 +1,380 @@
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+  }
+
+  abstract getProtocolDetails(): ProtocolDetails
+
+  abstract addresses: {
+    veToken: string
+    underlyingToken: string
+    rewardToken: string
+    feeDistributor: string
+  }
+
+  abstract getRewardBalance(input: GetPositionsInput): Promise<bigint>
+
+  abstract getLockedDetails(
+    input: GetPositionsInput,
+  ): Promise<{ amount: bigint; end: bigint }>
+
+  /**
+   * Creates a token and isProtocol tag
+   */
+  createProtocolTokenMetadata(protocolToken: Erc20Metadata): Metadata[] {
+    return [
+      {
+        address: this.addresses.veToken,
+        tag_id: TagIds.isProtocolToken,
+        value: true,
+      },
+
+      {
+        address: protocolToken.address,
+        tag_id: TagIds.token,
+        value: protocolToken,
+      },
+    ]
+  }
+
+  /**
+   * Finds tokens which have isProtocolToken tag
+   */
+  async getProtocolTokenMetadata(): Promise<Erc20Metadata[]> {
+    const metadata = (await this.buildMetadata()) as Metadata[]
+
+    const protocolMetadata = metadata.filter(
+      (metadata) => metadata.tag_id === TagIds.isProtocolToken,
+    )
+
+    return protocolMetadata.reduce((acc: Erc20Metadata[], protocolToken) => {
+      const relatedTokenTag = metadata.find(
+        (metadata) =>
+          metadata.address === protocolToken.address &&
+          metadata.tag_id === TagIds.token,
+      )
+      if (relatedTokenTag) {
+        acc.push(relatedTokenTag.value as Erc20Metadata)
+      }
+      return acc
+    }, [])
+  }
+
+  /**
+   * Creates a token and underlying_token_link from the protocol token address to the underlying token address
+   */
+  createUnderlyingTokenMetadata(
+    parentTokenAddress: string,
+    underlyingToken: Erc20Metadata,
+  ): Metadata[] {
+    return [
+      {
+        address: parentTokenAddress,
+        tag_id: TagIds.underlyingTokenLink,
+        value: underlyingToken.address,
+      },
+
+      {
+        address: underlyingToken.address,
+        tag_id: TagIds.token,
+        value: underlyingToken,
+      },
+    ]
+  }
+
+  /**
+   * Finds tokens which are linked to the parent token address
+   */
+  async getUnderlyingTokenMetadata(
+    parentTokenAddress: string,
+  ): Promise<Erc20Metadata> {
+    const metadatas = (await this.buildMetadata()) as Metadata[]
+    const link = metadatas.find(
+      (metadata) =>
+        metadata.tag_id === TagIds.underlyingTokenLink &&
+        metadata.address === parentTokenAddress,
+    )
+    const underlyingToken = metadatas.find(
+      (metadata) =>
+        metadata.address === link?.value && metadata.tag_id === TagIds.token,
+    )
+
+    return underlyingToken?.value as Erc20Metadata
+  }
+
+  /**
+   * Creates a reward manager tag
+   */
+  createRewardManagerMetadata(address: string): Metadata {
+    return {
+      address: address,
+      tag_id: TagIds.isRewardManager,
+      value: true,
+    }
+  }
+
```

Comment:
Left the utils in this adapter for now......if we think the solution has some appeal these could live in a MetadataBuilder class